akka-sample-distributed-workers-scala/src/main/scala/worker/WorkManager.scala

```diff
@@ -19,169 +19,124 @@ import worker.WorkState.WorkerFailed
-  final case class WorkerState(ref: ActorRef[WorkerCommand], status: WorkerStatus)
+//  sealed trait WorkerStatus
+//  case object Idle extends WorkerStatus
+//  final case class Busy(workId: String, deadline: Deadline) extends WorkerStatus
+//  final case class WorkerState(ref: ActorRef[WorkerCommand], status: WorkerStatus)
 
-  sealed trait Command extends CborSerializable
-  private final case class UpdatedWorkers(workers: Receptionist.Listing) extends Command
 
   // Messages from Workers
-  final case class WorkerRequestsWork(workerId: String, replyTo: ActorRef[WorkerCommand]) extends Command
-  final case class WorkIsDone(workerId: String, workId: String, result: Any, replyTo: ActorRef[WorkerCommand])
-      extends Command
-  final case class WorkFailed(worker: ActorRef[WorkerCommand], workId: String) extends Command
+ // final case class WorkerRequestsWork(workerId: String, replyTo: ActorRef[WorkerCommand]) extends Command
+  //final case class WorkIsDone(workerId: String, workId: String, result: Any, replyTo: ActorRef[WorkerCommand])
+  //    extends Command
+  //final case class WorkFailed(worker: ActorRef[WorkerCommand], workId: String) extends Command
 
   // Responses to requests from workers
   sealed trait WorkerCommand
   final case class DoWork(work: Work) extends WorkerCommand
-  final case class WorkAck(id: String) extends WorkerCommand
-  case object WorkAvailable extends WorkerCommand
 
-  // External commands
+  sealed trait Command extends CborSerializable
   final case class SubmitWork(work: Work, replyTo: ActorRef[WorkManager.Ack]) extends Command
+  private case class RequestNextWrapper(ask: RequestNext[WorkerCommand]) extends Command
+  final case class WorkIsDone(id: String) extends Command
+  final case class WorkFailed(id: String, t: Throwable) extends Command
+  private final case class TryStartWork() extends Command
 
   def apply(workTimeout: FiniteDuration): Behavior[Command] =
     Behaviors.setup { ctx =>
       Behaviors.withTimers { timers =>
         // FIXME use typed pub sub once https://github.com/akka/akka/issues/26338 is done
+
+        implicit val timeout = Timeout(5.seconds)
         val mediator = DistributedPubSub(ctx.system.toClassic).mediator
+        val producerController = ctx.spawn(WorkPullingProducerController[WorkerCommand]("work-manager", ManagerServiceKey, None), "producer-controller")
+        val requestNextAdapter = ctx.messageAdapter(RequestNextWrapper)
+        producerController ! WorkPullingProducerController.Start(requestNextAdapter)
 
-        // the set of available workers is not event sourced as it depends on the current set of workers
-        var workers = Map[ActorRef[WorkerCommand], WorkerState]()
+        var requestNext = Queue[RequestNext[WorkerCommand]]()
 
-        def notifyWorkers(workState: WorkState): Unit =
+        def tryStartWork(workState: WorkState): Effect[WorkDomainEvent, WorkState] = {
```

Comment:
Shouldn't there be something for `RecoveryCompleted` that would resend the `pendingWork` to the WorkPullingProducerController?

Reply:
Yes. This will have to wait for demand from the WPPC. I wonder if we should allow something similar to sharding where we can send more than one message?

Reply:
It is the workInProgress that we need to resent, the pendingWork hasn't been sent yet and will when demand is received. We can have an event that puts work in progress back into the pending queue